[quant][observer] Refactor MinMax observer

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23902 [quant][observer] Refactor MinMax observer**

Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
[quant][observer] Refactor MinMax observer

Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -1,50 +1,62 @@
-        self.max_val = None
+        assert self.qscheme in (
+            torch.per_tensor_affine,
+            torch.per_tensor_symmetric,
```
we don't have much support for per_tensor_symmetric I think

raghuramank100 replied:
Lets continue to support it. This only affects calculate q params

jerryzh168 replied:
hmm, theoretically we need to pass on the qscheme to quantized tensor, it works only because we always call `quantize_linear` when we convert quantized modules right now. Or should we always represent per_tensor_symmetric using per_tensor_affine quantization in backend?

raghuramank100 replied:
We need a way for backends to know that zero-point is zero (for uint8) and 128( for int8). Perhaps use the same backend and have a field to state that zero-point maps to zero/128?

hx89 replied:
A n00b question, what's the difference between per_tensor_affine and per_tensor_symmetric here?

jerryzh168 replied:
@raghuramank100 what do you mean? if it is symmetric quantization, don't we have separate implementation of quantized kernels?

@hx89 per_tensor_symmetric only has scale as an argument and the float_min/float_max are always symmetric, please see: https://arxiv.org/pdf/1806.08342.pdf

jerryzh168 reviewed (approved):
LGTM w/ minor comments, I think we only allow `torch.per_tensor_affine` qscheme right now

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -1,50 +1,62 @@
+
+        # extend min/max values to include 0 to meet the requirement that 0 is
+        # exactly repsentable
+        min_val = min(min_val, 0.0)
```
Dont you need torch.min here?

hx89 replied:
I saw torch.min is to get the min value of a tensor, here is to get the min of two numbers, do we need torch.min in this case?

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -330,6 +330,18 @@ def test_fuse_module_eval(self):
+        self.assertEqual(myobs.max_val, 8.0)
+        qparams = myobs.calculate_qparams()
+        self.assertAlmostEqual(qparams[0].item(), 0.0313725, delta=1e-5)
+        self.assertEqual(qparams[1].item(), 0.0)
```
Please add tests for per_tensor_affine and for qint8/quint8 combinations.

raghuramank100 reviewed (changes requested):
Please see comments, thanks

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

hx89 commented:
@raghuramank100 , I have replied your comment, could you please take a look?

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -1,52 +1,63 @@
+        assert self.dtype in (
+            torch.qint8,
+            torch.quint8,
+        ), "Default Observer only works for qint8 and quint data type"
```
NIT: quint8

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -607,11 +606,12 @@ def checkQuantized(model):
-    def test_observer(self, qdtype, qscheme):
-        myobs = Observer(dtype=qdtype, qscheme=qscheme)
+    def test_minmax_observer(self, qdtype, qscheme):
+        myobs = MinMaxObserver(dtype=qdtype, qscheme=qscheme)
```
Have you rebased to tip? Please verify that you have changes from https://github.com/pytorch/pytorch/pull/24996/files and ensure that observer is scriptable.

raghuramank100 reviewed (approved):
Please ensure that you rebased to tip and all observer tests pass  including the ones in 
https://github.com/pytorch/pytorch/pull/24996/files

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)

Commit:
Update on "[quant][observer] Refactor MinMax observer"


Copied from Daya's diff in pytorch/pytorch #23191

Refactor MinMax observer and create the base observer class to prepare for future observers such as histogram observer.

Differential Revision: [D16663221](https://our.internmc.facebook.com/intern/diff/D16663221/)
